### PR TITLE
upgrades to webpack 2.2

### PR DESCRIPTION
**Pull request by gabeweaver**
# Why
https://github.com/spartansystems/booster-kit-react/issues/25 

# What
* Updates dependencies
* Migrates webpack config to be inline with webpack 2.2
* Breaks out entry, output, and options into their own config files
* Removes lots of global configs and places them within their
respective webpack config files
* Configures all webpack configs to support passing in an env to build
a new config object from common params plus the environment params
* Updates feature specs and nightmare config so browser actions are
decoupled from feature tests and can be stacked onto of one another for
more configurable, extensible feature tests.
* Updates package.json scripts to use better-npm-run shorthand
* Adds yarn.lock file to version control
* Removes yarn clean file as it’s not safe for dependencies during
compilation

### CODE REVIEW

File: build/webpack.styles.js [hunk 1]
```diff
@@ -1,35 +1,61 @@
...
+        style,
+        ...cssOptions
       ]
-    })
-  }
+    }, {
+      test: /\.scss$/,
+      use: [
+        style,
+        ...sassOptions
```
Comment (bbrock25) [inline build/webpack.styles.js]: why is this different in development and test?

File: package.json [hunk 2]
```diff
@@ -10,25 +10,25 @@
   },
   "scripts": {
     "analyze": "flow",
-    "clean": "rimraf dist",
+    "build": "bnr build",
```
Comment (bbrock25) [inline package.json]: what is bnr?

File: tests/features/actions/visit.js [hunk 3]
```diff
@@ -0,0 +1,12 @@
+import nightmare from 'nightmare'
+import url from 'url'
+
+import { options, url as BASE_URL } from '../browser'
+
+export default function (path = '', query = {}) {
+  const location = url.resolve(BASE_URL, path)
+  return (
+    nightmare(options)
+      .goto(location)
```
Comment (bbrock25) [inline tests/features/actions/visit.js]: is this asserting anything?

File: package.json (hunk 2 above)
Comment (gabeweaver) [inline package.json]: better-npm-run -> https://github.com/benoror/better-npm-run

File: tests/features/actions/visit.js (hunk 3 above)
Comment (gabeweaver) [inline tests/features/actions/visit.js]: no its just a browser action. The way i've configured the feature tests is to destructure the browser actions as much as possible from the actual tests. Doing it this way allows the feature tests to be composed of many different browser actions (each spawning their own nightmare instance...thus avoiding the issues noted [here](https://github.com/rosshinkley/nightmare-examples/blob/master/docs/common-pitfalls/async-operations-loops.md)

Because we are using the async/await syntax, we can decompose all the steps into their own actions, each returning a promise, then we can await them, then do the assertions on the output from the resolved (or rejected) promise. 

This also allows for actions to be used across multiple feature tests so we don't write duplicate browser steps.

File: build/webpack.styles.js (hunk 1 above)
Comment (gabeweaver) [inline build/webpack.styles.js]: you mean development and production? In development, we don't want to extract all the styles because it adds unnecessary overhead and build time on each recompile...so when we're in development, we just want to use the style-loader to inject all the styles into the <head> of the HTML file. This is exponentially faster.

For production, eventually we want to get to the point of chunking our CSS down to individual CSS files and only loading the CSS we need on a given route....or we may opt to use inline JS styles - which if we go that direction - would enable us to remove the extract-text-webpack-plugin entirely since we wouldn't need to compile css files for production.

Comment (bbrock25) [inline build/webpack.styles.js]: yes, i misread that, thought that `test` was another key. Still not totally used to this webpack syntax...

File: tests/features/actions/visit.js (hunk 3 above)
Comment (bbrock25) [inline tests/features/actions/visit.js]: sounds good. I hope it works out ok, these things get a bit finicky.

File: package.json (hunk 2 above)
Comment (bbrock25) [inline package.json]: also, glad to have that clean script out. not sure if you've done it already, but having those test configs dynamically built is really hard to debug. I'd prefer duplication for those situations.
